wp/lib/bap_wp/src/precondition.ml

```diff
@@ -1397,47 +1401,36 @@ let in_loop (loop : Graphs.Ir.Node.t group) (tid : Tid.t) : bool =
    - for each node in the SCC, find its parent in the dominator tree
    - if the parent node is not in the original SCC, it is an exit node.
 
-   The exit node of a loop is the first node outside of the SCC in the control
-   flow graph. This should be the node the loop header points to that isn't a
-   part of the loop.
+   The exit node of a loop is a node outside of the SCC in the control flow
+   graph. This node will always be visited when exiting a loop.
 *)
-let loop_exit (node : Graphs.Ir.Node.t) (graph : Graphs.Ir.t)
+let loop_exit (loop : Graphs.Ir.Node.t group) (graph : Graphs.Ir.t)
   : Graphs.Ir.Node.t option =
-  let module Node = Graphs.Ir.Node in
-  let scc = Graphlib.strong_components (module Graphs.Ir) graph in
-  let* group = Partition.group scc node in
   let* leaf = Seq.find (Graphlib.postorder_traverse (module Graphs.Ir) graph)
-      ~f:(fun n -> Seq.is_empty (Node.succs n graph)) in
+      ~f:(fun n -> Seq.is_empty (Graphs.Ir.Node.succs n graph)) in
   let dom_tree = Graphlib.dominators (module Graphs.Ir) ~rev:true graph leaf in
-  Seq.find_map (Group.enum group) ~f:(fun n ->
+  Seq.find_map (Group.enum loop) ~f:(fun n ->
       let* parent = Tree.parent dom_tree n in
-      if Group.mem group parent then
+      if Group.mem loop parent then
         None
       else
         Some parent)
 
 (* Gets the precondition of the exit node of a loop. *)
 let exit_pre (env : Env.t) (post : Constr.t) (node : Graphs.Ir.Node.t)
```

Comment:
👍

Reply:
Though I guess we could do this *as we hit the nodes* during the loop analysis, but I'm not sure that would improve the code.

Reply:
Are you suggesting a hook in `visit_graph` where we calculate the parent as we enter the node, and then if it is a parent in the post-dominator tree, we recursively call `visit_graph` to update the env?

Reply:
Yes, that's what I'm suggesting, as maybe a `jmp_spec` or whatever we call it.

Again: not sure that would make the code cleaner in any way, so I wouldn't sweat it.